python/sqlflow_submitter/xgboost/feature_column.py

```diff
@@ -0,0 +1,243 @@
+        self.default_value = default_value
+
+    def _set_field_names(self, field_names):
+        CategoricalColumnTransformer._set_field_names(self, field_names)
+        self.column_idx = self.field_names.index(self.key)
+
+    def get_column_names(self):
+        return [self.key]
+
+    def __call__(self, inputs):
+        def transform_fn(slot_value):
+            invalid_index = slot_value < 0 or slot_value >= self.num_buckets
+            if any(invalid_index):
+                if self.default_value is not None:
+                    slot_value[invalid_index] = self.default_value
+                else:
+                    raise ValueError(
+                        'The categorical value of column {} out of range [0, {})'
+                        .format(self.field_names[self.column_idx],
+                                self.num_buckets))
+            return slot_value
+
+        return apply_transform_on_value(inputs[self.column_idx], transform_fn)
+
+
+def categorical_column_with_identity(key, num_buckets, default_value=None):
+    return CategoricalColumnWithIdentityTransformer(key, num_buckets,
+                                                    default_value)
+
+
+class CategoricalColumnWithVocabularyList(CategoricalColumnTransformer):
+    def __init__(self, key, vocabulary_list):
+        self.key = key
+        self.vocabulary_list = vocabulary_list
+
+    def _set_field_names(self, field_names):
+        CategoricalColumnTransformer._set_field_names(self, field_names)
+        self.column_idx = self.field_names.index(self.key)
+
+    def get_column_names(self):
+        return [self.key]
+
+    def __call__(self, inputs):
+        def transform_fn(slot_value):
+            if isinstance(slot_value, np.ndarray):
+                output = np.ndarray(slot_value.shape)
+                for i in six.moves.range(slot_value.size):
+                    output[i] = self.vocabulary_list.index(slot_value[i])
+            else:
+                output = self.vocabulary_list.index(slot_value)
+
+            return output
+
+        return apply_transform_on_value(inputs[self.column_idx], transform_fn)
+
+
+def categorical_column_with_vocabulary_list(key, vocabulary_list):
+    return CategoricalColumnWithVocabularyList(key, vocabulary_list)
+
+
+class CategoricalColumnWithHashBucketTransformer(CategoricalColumnTransformer):
+    def __init__(self, key, hash_bucket_size, dtype='string'):
+        self.key = key
+        self.hash_bucket_size = hash_bucket_size
+        self.dtype = dtype
+
+    def _set_field_names(self, field_names):
+        CategoricalColumnTransformer._set_field_names(self, field_names)
+        self.column_idx = self.field_names.index(self.key)
+
+    def get_column_names(self):
+        return [self.key]
+
+    def __call__(self, inputs):
+        def transform_fn(slot_value):
+            if isinstance(slot_value, np.ndarray):
+                output = np.ndarray(slot_value.shape)
```

Comment:
Maybe `output = [hashing(item) for item in slot_value]` is cleaner.

Reply:
It may not work. Python builtin function `hash` does not support `numpy.ndarray` and for-range iteration of the `numpy.ndarray` would yield each `numpy.ndarray` whose dimension is one less than the original one.